### ISSUE TITLE
Totem Complains About Missing Codecs For All Videos

### ISSUE THREAD

**Issue by Saroufim** [flathub/org.gnome.Totem#1]
Totem doesn't play any video at all, that includes Apple trailers and YouTube videos as well as local ones. If installed through dnf, it can play them all. 

Here's the relevant output:

```
Gtk-Message: Failed to load module "pk-gtk-module"
Gtk-Message: Failed to load module "pk-gtk-module"

(totem:2): Tracker-WARNING **: tracker-backend.vala:211: Falling back to bus backend, the direct backend failed to initialize: Locale change detected (DB:unknown, User/App:es_ES.utf8)

** Message: Missing plugin: gstreamer|1.0|totem|decodificador MPEG-4 AAC|decoder-audio/mpeg, mpegversion=(int)4, level=(string)2, base-profile=(string)lc, profile=(string)lc (decodificador MPEG-4 AAC)
** Message: Missing plugin: gstreamer|1.0|totem|decodificador H.264 (Main Profile)|decoder-video/x-h264, level=(string)3, profile=(string)main, interlace-mode=(string)progressive (decodificador H.264 (Main Profile))
** Message: Automatic missing codec installation not supported (helper script missing)
** Message: Missing plugin: gstreamer|1.0|totem|decodificador MPEG-4 Video (Simple Profile)|decoder-video/mpeg, mpegversion=(int)4, systemstream=(boolean)false, profile=(string)simple, level=(string)3 (decodificador MPEG-4 Video (Simple Profile))
** Message: Automatic missing codec installation not supported (helper script missing)
** Message: Missing plugin: gstreamer|1.0|totem|decodificador MPEG-4 AAC|decoder-audio/mpeg, mpegversion=(int)4, level=(string)2, base-profile=(string)lc, profile=(string)lc (decodificador MPEG-4 AAC)
** Message: Missing plugin: gstreamer|1.0|totem|decodificador H.264 (High Profile)|decoder-video/x-h264, level=(string)3.1, profile=(string)high, interlace-mode=(string)progressive (decodificador H.264 (High Profile))
** Message: Automatic missing codec installation not supported (helper script missing)
** Message: Missing plugin: gstreamer|1.0|totem|decodificador MPEG-4 AAC|decoder-audio/mpeg, mpegversion=(int)4, level=(string)2, base-profile=(string)lc, profile=(string)lc (decodificador MPEG-4 AAC)
** Message: Missing plugin: gstreamer|1.0|totem|decodificador H.264 (Main Profile)|decoder-video/x-h264, level=(string)3, profile=(string)main, colorimetry=(string)bt709, interlace-mode=(string)progressive (decodificador H.264 (Main Profile))
** Message: Automatic missing codec installation not supported (helper script missing)
** Message: Missing plugin: gstreamer|1.0|totem|decodificador text/html|decoder-text/html (decodificador text/html)
** Message: Automatic missing codec installation not supported (helper script missing)

(totem:2): Totem-WARNING **: Failed to reset the playback rate to 1.0
```

I've already got org.freedesktop.BaseSdk/x86_64/1.6, org.freedesktop.Sdk/x86_64/1.6,  org.freedesktop.Platform.VAAPI.Intel/x86_64/1.6, and the nvidia drivers installed. Supposedly these should hook into all the requirements that Totem needs from the host system to play videos. 

NB: Totem 'plays' certain .avi videos where the playback starts and the audio is there but the video remains completely blank.

**Comment by TingPing** [comment]
Well AAC, MPEG-4, and H.264 all make sense, none of those codecs are shipped in the runtime (we can bundle those here though).

Youtube in theory should be using webm/vp9/opus though and should work fine.

**Comment by nedrichards** [comment]
Sadly not everything on youtube is vp9/webm - but that that isn't h264 and aac cover, so it's not an extra consideration.

I assume that this is an NVidia system? Otherwise if you have VAAPI etc. then I'd expect that to work with hardware accelerated codecs. We might need to consider the NVidia acceleration a bit more.

**Comment by Saroufim** [comment]
Yeah, I've got the NVIDIA drivers installed on my host system as well as via flathub.

**Comment by hadess** [comment]
I'll look into adding more codecs out-of-the-box. The Flatpak was migrated from the GNOME builder which didn't allow me to ship codecs with patents applicable in the US. That's not the case in Flathub.

> NB: Totem 'plays' certain .avi videos where the playback starts and the audio is there but the video remains completely blank.

Please file a separate upstream bug about this. I believe that there's a bug in totem or GStreamer for that problem (totem is not supposed to play videos when it can't decode the video...)

**Comment by hadess** [comment]
> (helper script missing)

This would require a portal to fix.

> ** Message: Missing plugin: gstreamer|1.0|totem|decodificador text/html|decoder-text/html (decodificador text/html)

This is probably a quvi script missing.

**Comment by TingPing** [comment]
@hadess I was planning on just adding `gst-libav`.

**Comment by hadess** [comment]
The Fedora package also depends on `gstreamer1-plugins-ugly-free`. Not all the plugins in there are useful:
http://pkgs.fedoraproject.org/rpms/gstreamer1-plugins-ugly-free/blob/master/f/gstreamer1-plugins-ugly-free.spec#_137

But some are still necessary like the first 3 of the external plugins, and we'd also need either the mad plugin or the mpg123 one. I'd prefer the mad one.

**Comment by TingPing** [comment]
> and we'd also need either the mad plugin or the mpg123 one. I'd prefer the mad one.

As I understand it `mpg123` is better in every way, gstreamer even moved it to `good` next major version.

**Comment by hadess** [comment]
No, it's better in terms of licensing (LGPL vs. MAD's GPL). It's worse in terms of what broken MP3s you can throw at it, and the quality of the output.

**Comment by ahayzen** [comment]
I am also suffering from
```
** Message: Missing plugin: gstreamer|1.0|totem|MPEG-4 AAC decoder|decoder-audio/mpeg, mpegversion=(int)4, level=(string)2, base-profile=(string)lc, profile=(string)lc (MPEG-4 AAC decoder)
** Message: Missing plugin: gstreamer|1.0|totem|H.264 (High Profile) decoder|decoder-video/x-h264, level=(string)3, profile=(string)high, interlace-mode=(string)progressive (H.264 (High Profile) decoder)
```

I thought that maybe installing `org.freedesktop.Platform.ffmpeg` would help (as this has aac and h264), but I already had it installed. Does Totem just need to take advantage of this extension and allow its use ? (I see the extension was only created recently) or does Totem need other changes ?

**Comment by ramcq** [comment]
If something in the stack provides gst-libav (even the base runtime could), then Totem will be able to benefit from the more capable ffmpeg as shipped in the extension. Seems like @tsaunier is looking into it.

With my Endless hat on, I'd like it if Totem does want to use any even more "exciting" codecs that they could still be provided in an extension either to the runtime (which has a GST plugin extension point) or to Totem itself. That would allow redistribution in/from/to the US of the base runtimes/flatpaks without the codec extensions.

**Comment by DudleyDursley** [comment]
There is a similar issue [(Missing codecs like libav? #234)](https://github.com/flathub/flathub/issues/234) regarding flathub-flatpaks in general that has been closed because the freedesktop-sdk 18.08 now includes gst-libav. Does that mean that "org.freedesktop.Platform.html5-codecs/x86_64/18.08" is supposed to provide these codecs or do the GNOME runtimes require their own codec-extensions?

**Comment by TingPing** [comment]
@DudleyDursley Once this is updated to GNOME 3.30 it should inherit those freedesktop changes.

**Comment by ramcq** [comment]
The 18.08 html5-codecs (as the name suggests) extension is less extensive than the 1.6 ffmpeg extension - the original goal of the ffmpeg extension was to add just H.264, AAC as required for playing most HTML5 .mov content - someone else helpfully added all the codecs under the sun. The libav inside the 18.08 SDK doesn't have all the "exciting" codecs in - so the right approach for Totem would be to add a .Codecs extension (see Pitivi or Audacity) with a more populated libav, or potentially arrange for a separate codecs extension (under the https://gitlab.com/freedesktop-sdk/freedesktop-sdk/blob/18.08/elements/flatpak-images/platform.bst#L41 extension point).

**Comment by DudleyDursley** [comment]
Just tested the new release: 3.30 isn't able to play mp4 video from Youtube, so even H.264/MPEG-4 AVC isn't working. It plays the AAC audio though.

**Comment by hadess** [comment]
> Just tested the new release: 3.30 isn't able to play mp4 video from Youtube, so even H.264/MPEG-4 AVC isn't working. It plays the AAC audio though.

That's because nothing changed wrt codecs support with the updated SDK. It's still something I need to do but I haven't had time to actually do it yet.

**Comment by ramcq** [comment]
The HTML5 (AAC + H.264) extension is also broken atm - https://gitlab.com/freedesktop-sdk/freedesktop-sdk/issues/430

**Comment by hadess** [comment]
> The HTML5 (AAC + H.264) extension is also broken atm - https://gitlab.com/freedesktop-sdk/freedesktop-sdk/issues/430

We'll probably want to have another extension with more formats supported, so that more niche (or old) formats also work.

**Comment by ramcq** [comment]
Aye indeed. But if the built-in libgst-av would detect changes to libavcodec.so on the path, this extension could just include a more featureful libavcodec.

**Comment by hadess** [comment]
> Aye indeed. But if the built-in libgst-av would detect changes to libavcodec.so on the path, this extension could just include a more featureful libavcodec.

Right, but we also need the more featureful GStreamer demuxers for each of the formats we want to support, so that's something that needs doing.
